### ISSUE TITLE
[crypto, otbn] Speeding up Montgomery constant calculations

### ISSUE THREAD

**Issue by ballifatih** [lowRISC/opentitan#16558]
@JPulkus has proposed some ideas during our crypto meeting for speeding up the algorithm that we use to compute some constants for Montgomery arithmetic, and I think we can use this PR to have a more broader discussion on that.

I think it will take some time for me to get the full context here and also ramp up on OTBN, but here is my understanding so far:
  * The so-called algorithm computes `R^2 mod P` and `P^{-1} mod R`, where `R = 2^256` and `P` is typically a large prime (group order of field prime).
  * These two constants (as they are _constants_) do not need to be computed for every multiplication/addition/inversion. I would assume that we also compute these constants once at the beginning of a very long operation such as ECDSA signing/key generation/verification, which invokes many Montgomery multiplication+reduction operations. Therefore, this algorithm is not on the critical path in the sense that even significant speed-up improvements have small gain in the total speed-up. 

I suppose if we decide to replace the existing algorithm, then we would probably want to have a written algorithm that is easier to understand and analyze (at least for the doc purposes). So, this issue can also be useful for that purpose.

I am self-assigning this, but I have some higher priority tasks in my task scheduler at the moment, so it might take me a while to actually start working on this. Also a related PR: #14944

Please feel free to correct me or add more context @jadephilipoom, @felixmiller, @johannheyszl, @JPulkus

**Comment by ballifatih** [comment]
I spent some time digesting [the algorithm that uses Newton's method](http://marc-b-reynolds.github.io/math/2017/09/18/ModInverse.html) to more quickly compute the multiplicative inverse of an odd number modulo 2^k. 

Let me briefly explain the difference between the plain algorithm (the one implemented in our repo), and the faster algorithm (with Newton's method). Both algorithms compute the inverse of `x` modulo `2^k`, i.e. the algorithms return `y` such that `xy = 1 mod 2^k`. In our case, `k = 256`.

In the plain algorithm (which is already implemented), this computation is done in a bit-by-bit fashion. Namely, we start by finding the least significant bit of `y` and then gradually reconstruct more significant bits  of `y`. Therefore, the loop consists of 256 iterations, and each iteration requires 1 comparison (= subtraction) and 1 multiplication. 

In the faster algorithm, we similarly start by reconstructing least significant bits of `y`. However, in this algorithm, instead of reconstructing 1 bit of `y`, we double the amount of known bits of `y`. In other words, if we already know `r` least significant bits of y in a round, we then derive the following more significant `r` bits. Therefore, if we start with 3-bit initial value (which is already `x mod 8`), then we can reconstruct (6, 12, 24, 48, ...) bits in the following iterations. All in all, the loop takes around 7 iterations to complete for 256-bit values. Each iteration is slightly more costly than the former algorithm: 2 multiplications and 1 additions.

Therefore, I agree with @JPulkus that there is a good amount of gain at the algorithmic level. This algorithm is also elegant.

However, while this looks like a good improvement at the algorithmic level, I am still not sure we have enough reasons to prioritize implementing this algorithm. My point of view is rather pragmatic here, rather than judging both algorithms by their technical merits.

In order to understand the impact of _any_ speed-up in this algorithm, I tried to find an example computation cost from the existing code in our repo.
 
To understand the impact of a speed-up in this algorithm, let me try to somewhat capture the cost of the current plain algorithm (Algorithm 3 of https://eprint.iacr.org/2017/1057),  in a larger context.

 * RSA-3072 signature verification: The `compute_m0_inv` part takes about somewhere between %5 - %10 of the total cost, in particular the `compute_m0_inv` algorithm takes around 9368 OTBN clock cycles. Here the relative cost of computing the inverse is a bit inflated, becase the exponent in the signature verification is small (e = 65536), which makes the rest of the computation small.
 * I am not able to provide similar figures for ECDSA at the moment, because we do not have readily available code from which I can simply extract the cost. However, the exponents in elliptic-curves are usually around the same order with the log group size. So, smaller fraction of cost goes to constant computation.
 * @jadephilipoom has also pointed out that RSA signature verification is not on the critical path. OTBN call here is invoked in asynchronous manner, so it does not block the execution path on Ibex.

In short, I should drop this algorithm from my task list, for the sake of prioritizing other incomplete tasks.

I am currently un-assinging myself, in case anyone else has time to work on this. Thanks for sharing this idea @JPulkus.

**Comment by johannheyszl** [comment]
Thanks @ballifatih and @JPulkus. This is interesting and elegant. 
However, due to the small impact on overall performance, we need to put it into the backlog for now and revisit at a later time.

**Comment by ballifatih** [comment]
Let me add two more points from our offline discussions before I forget/lose them in my mailbox:

- What I missed during last week's discussion was that @JPulkus suggests this idea also helps to prevent the burden of storing this constant somewhere in NVM (or transmitting it to device). It might be an alternative solution, if we want to avoid storing these values in NVM. 
- At the moment, minimizing storage in NVM is not one of our immediate concerns (small discussion with @jadephilipoom). Also this value is relatively small in comparison to RSA public key.
- #14944 is the implementation for the algorithm above and we do not need to implement it from scratch.

**Comment by johannheyszl** [comment]
thx.
Regarding storage: I had last assumed we either load the constant from NVM once or compute once, then store in OTBN SRAM? So load / transmit overhead is minimal?

**Comment by msfschaffner** [comment]
What is the status of this issue? 
Would this still be good to complete?

**Comment by felixmiller** [comment]
Afaik the Montgomery constant computation has been highly improved by @jadephilipoom in the meantime. Also afaik, we use Montgomery only for RSA (Barrett for anything ECC related).

**Comment by jadephilipoom** [comment]
Correct, the Montgomery constant computations only matter for RSA (we don't currently use Montgomery for ECC, but even if we did we could just precompute the constants because the modulus there is fixed at compile-time).

The Montgomery R^2 computation has already been improved. The other constant, `m0inv`, is still slow, and we could improve it with the algorithm from #14944. We could also store both constants along with the RSA key to speed up sign/verify/encrypt/decrypt operations; this is an implementation choice for the cryptolib that hasn't been fully settled. I'd guess storing `m0inv` at least will be worth it because it's relatively small.

One place where we will definitely run these constant computations, and repeatedly, is RSA key generation. For that purpose, it might be very good to have a faster computation of these constants. However, the rest of RSA key generation (particularly the primality test) is so slow that right now I don't think it would make a significant performance difference; there's lower-hanging fruit there.
